clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -349,7 +394,7 @@ public long timeToNextPoll(long now) {
 
         // update metadata (if needed) and keep track of the metadata used for assignment so that
         // we can check after rebalance completion whether anything has changed
-        client.ensureFreshMetadata();
+        if (!client.ensureFreshMetadata(Long.MAX_VALUE)) throw new TimeoutException();
```

Comment:
I guess it's tough to avoid blocking here. Maybe room for improvement in the future. All of this would be considerably easier if we moved the rebalance to the background thread. Another improvement for another time.